### ISSUE TITLE
Half Life Regression

### ISSUE THREAD

**Issue by alexanderquispe** [alexanderquispe/ForgettingCurve#8]
What are the functions?  - @josuecaldasv 
Finish the Gradient Descent. 
Deploy Model. 

Possible Extensions
* Lasso Reg
* Ridge
* Elastic Net

**Comment by josuecaldasv** [comment]
Las funciones son las siguientes:

1. **estimate_h_hat**: Permite estimar h_hat dados los valores de theta y x. Ecuación N. 2. 
2. **estimate_p_hat**: Permite estimar p_hat dados los valores de delta y h_hat. Ecuación N. 4.
3. **hh_loss_function**: Permite estimar el costo de la función tomando en cuenta el valor de p, p_hat, h_hat, delta y lambda. Ecuación N. 9. 
4. **gradient_partial**: Permite estimar el gradiente de la función de costo. Ecuación N. 10.
5. **adagrad_update**: Permite actualizar el valor de theta con cada iteración utilizando el algoritmo SGD adagrad. Ecuación N. 11.
6. **half_life_regression**: Modelo HLR que integra todas las funciones anteriores. 

La función **half_life_regression** se encuentra todavía en desarrollo.

**Comment by alexanderquispe** [comment]
AdaGrad
```

import numpy as np

class LogisticRegressionWithAdaGrad:
    def __init__(self, learning_rate=0.01, max_iter=1000, epsilon=1e-8):
        self.learning_rate = learning_rate
        self.max_iter = max_iter
        self.epsilon = epsilon  # small number to avoid division by zero

    def fit(self, X, y):
        m, n = X.shape
        self.weights = np.zeros(n)
        self.bias = 0
        self.G_square_sum = np.zeros(n)  # Sum of squares of gradients

        for _ in range(self.max_iter):
            model = 1 / (1 + np.exp(-(X.dot(self.weights) + self.bias)))
            error = model - y

            # Gradient w.r.t weights and bias
            gradient_w = X.T.dot(error) / m
            gradient_b = np.sum(error) / m

            # Update sum of squares of gradients
            self.G_square_sum += gradient_w ** 2

            # AdaGrad adjustment
            adjusted_lr = self.learning_rate / (np.sqrt(self.G_square_sum) + self.epsilon)

            # Update weights and bias
            self.weights -= adjusted_lr * gradient_w
            self.bias -= self.learning_rate * gradient_b

    def predict(self, X):
        model = 1 / (1 + np.exp(-(X.dot(self.weights) + self.bias)))
        return np.where(model >= 0.5, 1, 0)

```

**Comment by alexanderquispe** [comment]
@josuecaldasv to finish ,

- Work  AdaGrad 
- Confirm if we find the global or local minimum in Sparse Function with Regularization.

**Comment by alexanderquispe** [comment]
![image](https://github.com/alexanderquispe/ForgettingCurve/assets/27368860/368a888b-eea7-4677-88c3-a1fabdcd2a55)
![image](https://github.com/alexanderquispe/ForgettingCurve/assets/27368860/e45eb564-8984-48a0-a94b-f94ab47ffa32)
